qualtran/surface_code/quantum_error_correction_scheme_summary.py

```diff
@@ -0,0 +1,114 @@
+#  Copyright 2023 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+import abc
+import enum
+
+import numpy as np
+from attrs import field, frozen
+
+from qualtran.surface_code.physical_parameters import PhysicalParameters
+
+
+@frozen
+class QuantumErrorCorrectionSchemeSummary(abc.ABC):
+    r"""QuantumErrorCorrectionSchemeSummary represents a high level view of a QEC scheme.
+
+    QuantumErrorCorrectionSchemeSummary provides estimates for the logical error rate,
+    number of physical qubits and the logical time step given a code distance and
+    physical assumptions.
+
+    The logical error rate as a function of code distance $d$ and physical error rate $p$
+    is given by
+    $$
+    a \left ( \frac{p}{p^*}  \right )^\frac{d + 1}{2}
+    $$
+    Where $a$ is the error_rate_scaler and $p^*$ is the error_rate_threshold.
+
+    Note: The logical error-suppression factor $\Lambda = \frac{p^*}{p}$
+
+    Attributes:
+        error_rate_scaler: Logical error rate coefficient.
+        error_rate_threshold: Logical error rate threshold.
+        reference: source of the estimates.
+    """
+
+    error_rate_scaler: float = field(default=0.1, repr=lambda x: f'{x:g}')
+    error_rate_threshold: float = field(default=0.01, repr=lambda x: f'{x:g}')
+    reference: str | None = None
+
+    def logical_error_rate(
+        self, code_distance: int | np.ndarray, physical_error_rate: float | np.ndarray
+    ) -> float | np.ndarray:
+        """The logical error rate for the given code distance using this scheme."""
+        return self.error_rate_scaler * np.power(
+            physical_error_rate / self.error_rate_threshold, (code_distance + 1) / 2
+        )
+
+    @abc.abstractmethod
+    def physical_qubits(self, code_distance: int | np.ndarray) -> int | np.ndarray:
+        """The number of physical qubits for the given code distance using this scheme."""
+
+    @abc.abstractmethod
+    def logical_time_step(
+        self, code_distance: int | np.ndarray, physical_parameters: PhysicalParameters
+    ) -> float | np.ndarray:
+        """The duration of a logical time step for the given code distance using this scheme."""
+
+
+class GateBasedSurfaceCodeSource(enum.Enum):
+    arXiv12080928 = 0  # Folwer Model.
+    arXiv221107629 = 1  # Azure Model.
```

Comment:
this isn't the most user-friendly api. The thing the user enters should probably be the memorable thing; not the opaque identifier. As far as I can tell: this just modifies the formula for `logical_time_step` from 4t_gate + 1t_meas to 4_tgate + 2t_meas.
 - make this difference apparent in the documentation
 - you probably don't need an enum for that -- just take the parameter directly
 - why is there a discrepancy?
 - Isn't this sortof moot because the fowler model (IIRC) just assumes a logical_time_step of 1 microsecond and doesn't really care how that decomposes into gate and measurement times